coil-gif/src/nonAndroidMain/kotlin/coil3/gif/AnimatedSkiaImage.kt

```diff
@@ -0,0 +1,233 @@
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import org.jetbrains.skia.AnimationFrameInfo
+import org.jetbrains.skia.Bitmap
+import org.jetbrains.skia.Codec
+
+internal class AnimatedSkiaImage(
+    private val codec: Codec,
+    private val coroutineScope: CoroutineScope,
+    private val timeSource: TimeSource,
+    private val animatedTransformation: AnimatedTransformation? = null,
+    private val onAnimationStart: (() -> Unit)? = null,
+    private val onAnimationEnd: (() -> Unit)? = null,
+    bufferedFramesCount: Int,
+) : Image {
+
+    override val size: Long
+        get() {
+            var size = codec.imageInfo.computeMinByteSize().toLong()
+            if (size <= 0L) {
+                // Estimate 4 bytes per pixel.
+                size = 4L * codec.width * codec.height
+            }
+            return size.coerceAtLeast(0)
+        }
+
+    override val width: Int
+        get() = codec.width
+
+    override val height: Int
+        get() = codec.height
+
+    override val shareable: Boolean
+        get() = false
+
+    /**
+     * The number of times the animation must be played through.
+     */
+    val maxIterationCount: Int
+        get() = codec.repetitionCount + 1
+
+    /**
+     * The duration of each frame, in order, in milliseconds.
+     */
+    val frameDurationsMs: List<Int> by lazy {
+        codec.framesInfo.map { it.safeFrameDuration }
+    }
+
+    /**
+     * The total duration of a single iteration of the animation in milliseconds.
+     */
+    val singleIterationDurationMs: Int by lazy {
+        frameDurationsMs.sum()
+    }
+
+    /**
+     * A temporary [Bitmap] used to decode individual frames.
+     */
+    private val tempBitmap = Bitmap().apply { allocPixels(codec.imageInfo) }
+
+    /**
+     * An array of decoded frames.
+     *
+     * Initially load `bufferedFramesCount` frames; the rest will be loaded in the background.
+     */
+    private val frames = Array(codec.frameCount) { index ->
+        if (index in 0..<bufferedFramesCount) decodeFrame(index) else null
+    }
+
+    private var bufferFramesJob: Job? = null
+
+    /**
+     * A value that is incremented to force the image to be invalidated by Compose.
+     */
+    private var invalidateTick by mutableIntStateOf(0)
+
+    /**
+     * The time when the animation first started playing. Used to calculate the current frame to be drawn.
+     */
+    private var animationStartTime: TimeMark? = null
+
+    private var hasNotifiedAnimationEnd = false
+
+    override fun draw(canvas: Canvas) {
+        if (codec.frameCount == 0) {
+            return
+        }
+
+        if (codec.frameCount == 1) {
+            canvas.drawFrame(frameIndex = 0)
+            return
+        }
+
+        // Buffer the next frames in the background.
+        if (bufferFramesJob == null || bufferFramesJob?.isCancelled == true) {
+            bufferFramesJob = coroutineScope.launch(Dispatchers.Default) {
```

Comment:
Let's add an option for `decoderCoroutineContext` to this class' constructor. We can pass the value from `options.decoderCoroutineContext` from our Decoder.